streams/src/test/java/org/apache/kafka/streams/integration/OptimizedKTableIntegrationTest.java

```diff
@@ -0,0 +1,293 @@
+        final Properties producerProps = new Properties();
+        producerProps.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, cluster.bootstrapServers());
+        producerProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, IntegerSerializer.class);
+        producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, IntegerSerializer.class);
+
+        IntegrationTestUtils.produceKeyValuesSynchronously(
+            INPUT_TOPIC_NAME,
+            IntStream.range(start, endExclusive)
+                .mapToObj(i -> KeyValue.pair(key, i))
+                .collect(Collectors.toList()),
+            producerProps,
+            mockTime);
+    }
+
+    private void waitForClientsToEnterRunningState(final Collection<KafkaStreams> clients,
+                                                   final long time,
+                                                   final TimeUnit timeUnit) throws InterruptedException {
+
+        final long expectedEnd = System.currentTimeMillis() + timeUnit.toMillis(time);
+
+        clientStatesLock.lock();
+        try {
+            while (!clients.stream().allMatch(client -> clientStates.get(client) == State.RUNNING)) {
+                if (expectedEnd <= System.currentTimeMillis()) {
+                    assertThat("requested clients entered " + State.RUNNING + " in a timely manner", false);
+                }
+                final long millisRemaining = Math.max(1, expectedEnd - System.currentTimeMillis());
+                clientStateUpdate.await(millisRemaining, TimeUnit.MILLISECONDS);
+            }
+        } finally {
+            clientStatesLock.unlock();
+        }
+    }
+
+    private KafkaStreams createClient(final StreamsBuilder builder, final Properties config) {
+        final KafkaStreams client = new KafkaStreams(builder.build(config), config);
+        clientStatesLock.lock();
+        try {
+            clientStates.put(client, client.state());
+        } finally {
+            clientStatesLock.unlock();
+        }
+
+        client.setStateListener((newState, oldState) -> {
+            clientStatesLock.lock();
+            try {
+                clientStates.put(client, newState);
+                if (newState == State.RUNNING) {
+                    if (clientStates.values().stream().allMatch(state -> state == State.RUNNING)) {
+                        clientStateUpdate.signalAll();
+                    }
+                }
+            } finally {
+                clientStatesLock.unlock();
+            }
+        });
+        return client;
+    }
+
+    private StateRestoreListener createTrackingRestoreListener(final AtomicLong restoreStartOffset) {
+        return new StateRestoreListener() {
+            @Override
+            public void onRestoreStart(final TopicPartition topicPartition, final String storeName,
+                final long startingOffset, final long endingOffset) {
```

Comment:
nit: we prefer the following formatting (similar below)

```
public void onRestoreStart(final TopicPartition topicPartition,
                           final String storeName,
                           final long startingOffset,
                           final long endingOffset) {
```

Reply:
Thanks. I thought I had grabbed the Kafka code style for IDEA. Do we have a different one for Kafka Streams?

Reply:
To some extend yes -- we tried to convince others to adopt Streams code style but it's always hard to get people to agree -- for example, Streams enforced `final` wherever possible, while clients don't do that -- and brokers are written in Scala anyway...

We cannot enforce everything via checkstyle unfortunately.